File: baekjoon/algo_11062.cpp
#include<iostream>
#include<string.h>
#include<algorithm>

using namespace std;

int dp[1001][1001][2];
int cards[1001];

int solve(int turn, int x, int y)
{
    int& reference = dp[x][y][turn];
    
    if(x == y)
    {
        if(turn == 0)
            return cards[x];
        else
            return 0;
    }
        
    if(reference != -1)
        return reference;
    
    if(turn == 0)
        reference = max(solve(turn^1, x+1, y) + cards[x], solve(turn^1, x, y-1) + cards[y]);
    else
        reference = min(solve(turn^1, x+1, y), solve(turn^1, x, y-1));
    
    return reference;
}
    

int main()
{
    int T;
    cin >> T;
    
    while(T--)
    {
        int N;
        cin >> N;
        memset(dp, -1, sizeof(dp));
        
        for(int i = 0; i < N; i++)
            scanf("%d", &cards[i]);
        printf("%d\n", solve(0, 0, N-1));
    }
    
    return 0;
}

File: baekjoon/algo_11062.py
from sys import stdin

def solve(turn: bool, x:int, y:int) -> int:
    if x == y:
        if turn == 0:
            return cards[x] 
        else:
            return 0 
    
    if dp[x][y][turn] is not -1:
        return dp[x][y][turn]
    
    if turn == 0:
        dp[x][y][turn] = max(solve(not turn, x+1, y) + cards[x], solve(not turn, x, y-1) + cards[y])
    else:
        dp[x][y][turn] = min(solve(not turn, x+1, y), solve(not turn, x, y-1))
    
    return dp[x][y][turn]

T = int(stdin.readline())
for t in range(T):
    N = int(stdin.readline())
    cards =list(map(int, stdin.readline().split()))
    dp = [ [ [-1, -1] for _ in range(N)] for _ in range(N)]
    result = solve(0, 0, N-1)
    print(result)